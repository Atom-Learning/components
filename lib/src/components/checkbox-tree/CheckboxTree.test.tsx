import { render, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import * as React from 'react'

import { CheckboxTree } from '.'

const CheckboxTreeImplementation = (props) => {
  return (
    <CheckboxTree {...props}>
      <CheckboxTree.Item value="top level">Top level</CheckboxTree.Item>
      <CheckboxTree.Collapsible>
        <CheckboxTree.CollapsibleTrigger
          title="top level all item"
          label="Open: Top level trigger"
        >
          Trigger With Submenu
        </CheckboxTree.CollapsibleTrigger>
        <CheckboxTree.CollapsibleContent>
          <CheckboxTree.Item value={1}>1</CheckboxTree.Item>
          <CheckboxTree.Item value={2}>2</CheckboxTree.Item>
          <CheckboxTree.Collapsible>
            <CheckboxTree.CollapsibleTrigger
              label="Open: Second level trigger"
              title="all but nested a"
            >
              All but nested
            </CheckboxTree.CollapsibleTrigger>
            <CheckboxTree.CollapsibleContent>
              <CheckboxTree.Item value="1 nested">Nested: #1</CheckboxTree.Item>
              <CheckboxTree.Item value="2 nested">Nested: #2</CheckboxTree.Item>
            </CheckboxTree.CollapsibleContent>
          </CheckboxTree.Collapsible>
        </CheckboxTree.CollapsibleContent>
      </CheckboxTree.Collapsible>
    </CheckboxTree>
  )
}

const mockOnCheckedChange = jest.fn()

describe(`CheckboxTree component`, () => {
  it('renders a checkbox', async () => {
    const { container } = render(<CheckboxTreeImplementation />)

    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = render(<CheckboxTreeImplementation />)

    expect(await axe(container)).toHaveNoViolations()
  })

  it('respects defaultChecked', async () => {
    render(<CheckboxTreeImplementation defaultChecked={[1, '2 nested']} />)
    const checkboxItem1 = screen.getByTitle('1')
    expect(checkboxItem1).toBeChecked()

    const checkboxItem2 = screen.getByTitle('2')
    expect(checkboxItem2).not.toBeChecked()

    const checkboxItem2Nested = screen.getByTitle('2 nested')
    expect(checkboxItem2Nested).toBeChecked()
  })

  it('respects checked', async () => {
    render(<CheckboxTreeImplementation checked={[2, '1 nested']} />)
    const checkboxItem1 = screen.getByTitle('1')
    expect(checkboxItem1).not.toBeChecked()

    const checkboxItem2 = screen.getByTitle('2')
    expect(checkboxItem2).toBeChecked()

    const checkboxItem1Nested = screen.getByTitle('1 nested')
    expect(checkboxItem1Nested).toBeChecked()
  })

  it('calls onCheckedChange correctly when used uncontrolled', async () => {
    render(
      <CheckboxTreeImplementation
        defaultChecked={[1, '2 nested']}
        onCheckedChange={mockOnCheckedChange}
      />
    )
    screen.getByTitle(1).click()

    expect(mockOnCheckedChange).toHaveBeenCalledWith(['2 nested'])
  })

  it('calls onCheckedChange correctly when used controlled', async () => {
    render(
      <CheckboxTreeImplementation
        checked={[2, '1 nested']}
        onCheckedChange={mockOnCheckedChange}
      />
    )
    screen.getByTitle('1').click()

    expect(mockOnCheckedChange).toHaveBeenCalledWith([2, '1 nested', 1])
  })
})
