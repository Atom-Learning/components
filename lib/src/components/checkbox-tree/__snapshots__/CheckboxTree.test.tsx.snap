// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxTree component renders a checkbox 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-gkooaz {
    width: 100%;
    padding: 0;
    margin: 0;
    list-style: none;
  }

  .c-gkooaz .c-gkooaz {
    padding-left: var(--space-4);
  }

  .c-jiJsop:not(:first-child) {
    margin-top: var(--space-0);
  }

  .c-rbNxa {
    width: 100%;
    position: relative;
    min-height: var(--sizes-3);
    padding-left: var(--space-6);
  }

  .c-cFpcyF {
    max-width: 100%;
  }

  .c-jgUzPy {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translateX(-50%) translateY(-50%);
  }

  .c-dScekx {
    -webkit-appearance: none;
    appearance: none;
    position: relative;
    background-color: transparent;
    border: 1px solid var(--colors-tonal400);
    border-radius: 3px;
    color: white;
    cursor: pointer;
    height: var(--sizes-1);
    width: var(--sizes-1);
    padding: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    transition: all 50ms ease-out;
    flex-shrink: 0;
  }

  .c-dScekx[data-state="checked"] {
    background-color: var(--colors-primary800);
    border-color: var(--colors-primary800);
  }

  .c-dScekx[data-state="indeterminate"] {
    background-color: var(--colors-primary800);
    border-color: var(--colors-primary800);
  }

  .c-dScekx:focus {
    outline: 2px solid var(--colors-primary800);
    outline-offset: 1px;
  }

  .c-dScekx[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-dyvMgW {
    font-family: var(--fonts-body);
    font-weight: 400;
    margin: 0;
  }

  .c-dyvMgW > .c-dyvMgW:before,
  .c-dyvMgW > .c-dyvMgW:after {
    display: none;
  }

  .c-bWcumU {
    padding-top: var(--space-0);
    padding-bottom: var(--space-0);
    white-space: nowrap;
    overflow-x: hidden;
    text-overflow: ellipsis;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-hBEOJc {
    position: absolute;
    left: 0;
    top: 0;
    pointer-events: none;
  }

  .c-hBEOJc .c-hMRxhp {
    transition: transform 300ms;
  }

  .c-hBEOJc[data-state="closed"] > .c-hMRxhp {
    transform: rotate(-90deg);
  }

  .c-hBEOJc[data-state="open"] > .c-hMRxhp {
    transform: rotate(0deg);
  }

  .c-gsmDXe {
    cursor: pointer;
  }

  .c-rvbcR[data-state="closed"] {
    display: none;
  }

  .c-fGHEql {
    width: 100%;
  }
}

@media  {
  .c-dhzjXW-iTKOFX-direction-column {
    flex-direction: column;
  }

  .c-dhzjXW-jroWjL-align-center {
    align-items: center;
  }

  .c-dhzjXW-kaVYFn-gap-2 {
    gap: var(--space-2);
  }

  .c-dScekx-kupcZn-size-md {
    height: var(--sizes-1);
    width: var(--sizes-1);
  }

  .c-dyvMgW-gvmVBy-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
  }

  .c-dyvMgW-gvmVBy-size-md::before {
    content: '';
    margin-bottom: -0.3864em;
    display: table;
  }

  .c-dyvMgW-gvmVBy-size-md::after {
    content: '';
    margin-top: -0.3864em;
    display: table;
  }

  .c-dyvMgW-fmPQHp-noCapsize-true::before,
  .c-dyvMgW-fmPQHp-noCapsize-true::after {
    display: none !important;
  }

  .c-oUmPa-elrktp-size-sm {
    height: var(--sizes-3);
    width: var(--sizes-3);
  }

  .c-hMRxhp-cyeZeh-size-sm {
    height: var(--sizes-1);
    width: var(--sizes-1);
    stroke-width: 1.5;
  }
}

@media  {
  .c-oUmPa-jCxibG-cv {
    background: transparent;
    color: var(--colors-tonal300);
  }

  .c-oUmPa-jCxibG-cv:not(:disabled):hover,
  .c-oUmPa-jCxibG-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-jCxibG-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

<div>
  <ul
    class="c-dhzjXW c-gkooaz c-dhzjXW-iTKOFX-direction-column c-PJLV PJLV"
    role="tree"
  >
    <li
      class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
      role="treeitem"
    >
      <div
        class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
      >
        <label
          class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
        >
          <button
            aria-checked="false"
            class="c-dScekx c-dScekx-kupcZn-size-md"
            data-state="unchecked"
            role="checkbox"
            title="top level"
            type="button"
            value="on"
          />
          <span
            class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
          >
            Top level
          </span>
        </label>
      </div>
    </li>
    <li
      class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
      role="treeitem"
    >
      <div
        class="c-fGHEql PJLV"
        data-state="closed"
      >
        <div
          class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2 c-gsmDXe"
        >
          <button
            aria-controls="radix-1"
            aria-expanded="false"
            aria-label="Open: Top level trigger"
            class="c-oUmPa c-oUmPa-elrktp-size-sm c-oUmPa-jCxibG-cv c-hBEOJc"
            data-state="closed"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-cyeZeh-size-sm c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 10 12 16 18 10"
              />
            </svg>
          </button>
          <label
            class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
          >
            <button
              aria-checked="false"
              class="c-dScekx c-dScekx-kupcZn-size-md"
              data-state="unchecked"
              role="checkbox"
              title="top level all item"
              type="button"
              value="on"
            />
            <span
              class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
            >
              Trigger With Submenu
            </span>
          </label>
        </div>
        <ul
          class="c-dhzjXW c-gkooaz c-dhzjXW-iTKOFX-direction-column c-rvbcR PJLV"
          data-state="closed"
          id="radix-1"
          role="group"
          style="transition-duration: 0s; animation-name: none;"
        >
          <li
            class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
            role="treeitem"
          >
            <div
              class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
            >
              <label
                class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
              >
                <button
                  aria-checked="false"
                  class="c-dScekx c-dScekx-kupcZn-size-md"
                  data-state="unchecked"
                  role="checkbox"
                  title="1"
                  type="button"
                  value="on"
                />
                <span
                  class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
                >
                  1
                </span>
              </label>
            </div>
          </li>
          <li
            class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
            role="treeitem"
          >
            <div
              class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
            >
              <label
                class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
              >
                <button
                  aria-checked="false"
                  class="c-dScekx c-dScekx-kupcZn-size-md"
                  data-state="unchecked"
                  role="checkbox"
                  title="2"
                  type="button"
                  value="on"
                />
                <span
                  class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
                >
                  2
                </span>
              </label>
            </div>
          </li>
          <li
            class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
            role="treeitem"
          >
            <div
              class="c-fGHEql PJLV"
              data-state="closed"
            >
              <div
                class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2 c-gsmDXe"
              >
                <button
                  aria-controls="radix-0"
                  aria-expanded="false"
                  aria-label="Open: Second level trigger"
                  class="c-oUmPa c-oUmPa-elrktp-size-sm c-oUmPa-jCxibG-cv c-hBEOJc"
                  data-state="closed"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="c-hMRxhp c-hMRxhp-cyeZeh-size-sm c-hMRxhp-iPJLV-css"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polyline
                      points="6 10 12 16 18 10"
                    />
                  </svg>
                </button>
                <label
                  class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                >
                  <button
                    aria-checked="false"
                    class="c-dScekx c-dScekx-kupcZn-size-md"
                    data-state="unchecked"
                    role="checkbox"
                    title="all but nested a"
                    type="button"
                    value="on"
                  />
                  <span
                    class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
                  >
                    All but nested
                  </span>
                </label>
              </div>
              <ul
                class="c-dhzjXW c-gkooaz c-dhzjXW-iTKOFX-direction-column c-rvbcR PJLV"
                data-state="closed"
                id="radix-0"
                role="group"
                style="transition-duration: 0s; animation-name: none;"
              >
                <li
                  class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                  role="treeitem"
                >
                  <div
                    class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                  >
                    <label
                      class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                    >
                      <button
                        aria-checked="false"
                        class="c-dScekx c-dScekx-kupcZn-size-md"
                        data-state="unchecked"
                        role="checkbox"
                        title="1 nested"
                        type="button"
                        value="on"
                      />
                      <span
                        class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
                      >
                        Nested: #1
                      </span>
                    </label>
                  </div>
                </li>
                <li
                  class="c-dhzjXW c-jiJsop c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                  role="treeitem"
                >
                  <div
                    class="c-dhzjXW c-rbNxa c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                  >
                    <label
                      class="c-dhzjXW c-cFpcyF c-dhzjXW-jroWjL-align-center c-dhzjXW-kaVYFn-gap-2"
                    >
                      <button
                        aria-checked="false"
                        class="c-dScekx c-dScekx-kupcZn-size-md"
                        data-state="unchecked"
                        role="checkbox"
                        title="2 nested"
                        type="button"
                        value="on"
                      />
                      <span
                        class="c-dyvMgW c-dyvMgW-gvmVBy-size-md c-dyvMgW-fmPQHp-noCapsize-true c-bWcumU"
                      >
                        Nested: #2
                      </span>
                    </label>
                  </div>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>
`;
